modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+        Clock clock,
+        ThreadPool threadPool,
+        LongSupplier nowSupplier
+    ) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.clock = clock;
+        this.transportActionsDeduplicator = new ResultDeduplicator<>(threadPool.getThreadContext());
+        this.nowSupplier = nowSupplier;
+        this.scheduledJob = null;
+        this.pollInterval = DLM_POLL_INTERVAL_SETTING.get(settings);
+        this.defaultRolloverRequestSupplier = this::getDefaultRolloverRequest;
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(DLM_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        final boolean prevIsMaster = this.isMaster;
+        if (prevIsMaster != event.localNodeMaster()) {
+            this.isMaster = event.localNodeMaster();
+            if (this.isMaster) {
+                // we weren't the master, and now we are
+                onBecomeMaster();
+            } else {
+                // we were the master, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(DATA_LIFECYCLE_JOB_NAME)) {
+            logger.trace("DLM job triggered: " + event.getJobName() + ", " + event.getScheduledTime() + ", " + event.getTriggeredTime());
+            run(clusterService.state());
+        }
+    }
+
+    /**
+     * Iterates over the DLM managed data streams and executes the needed operations
+     * to satisfy the configured {@link org.elasticsearch.cluster.metadata.DataLifecycle}.
+     */
+    // default visibility for testing purposes
+    void run(ClusterState state) {
+        for (DataStream dataStream : state.metadata().dataStreams().values()) {
+            if (dataStream.getLifecycle() == null) {
+                continue;
+            }
+
+            IndexMetadata writeIndex = state.metadata().index(dataStream.getWriteIndex());
+            if (writeIndex != null && isManagedByDLM(dataStream, writeIndex)) {
+                RolloverRequest rolloverRequest = defaultRolloverRequestSupplier.apply(dataStream.getName());
+                transportActionsDeduplicator.executeOnce(
+                    rolloverRequest,
+                    ActionListener.noop(),
```

Comment:
Should we pass in some kind of listener for logging purposes so that DLM can log (at trace) that it's invoking rollover requests? Not sure if it'd be too much noise or whether it'd be useful, what do you think?

Reply:
++

I think we need a custom listener here that will collect the encountered errors. I am planning to add it as part of the next effort (follow-up PR) related to error reporting.